src/levanter/data/loader.py

```diff
@@ -223,25 +285,64 @@ def get_data(indices):
                 return raw_array
 
         gda_leaves = [
-            make_global_array_for_leaf(leaf_index, _batchified_shape(self.dl.Batch, item_leaf))
+            make_global_array_for_leaf(leaf_index, _batchified_shape(Batch, item_leaf))
             for leaf_index, item_leaf in enumerate(self.dl._ex_leaves)
         ]
         gda_tree = jax.tree.unflatten(self.dl._ex_structure, gda_leaves)
         return gda_tree
 
-    async def _do_retrieve_batch_of_batches(self, batch_numbers):
-        indices_for_this_batch_of_batches: list[int] = []
+    async def _do_retrieve_batch_of_batches(self, batch_numbers: list[int]) -> list[_Batch[Ex]]:
+        """
+        Retrieve the data for a batch of batches.
+
+        - For each batch number, we get the mapping from our local devices to the indices in the data store that we need
+          to load.
+        - We then load the data for all the devices for all the batch numbers.
+        - we then return each batch as a dictionary from device to the data for that device.
+        """
+        global_indices_for_each_batch: list[list[int]] = []
+        sizes = [self.dl.scheduler.batch_size_at_step(bn) for bn in batch_numbers]
+        offsets = []
+
         for bn in batch_numbers:
-            indices_this_batch = range(bn * self.dl.batch_size, (bn + 1) * self.dl.batch_size, 1)
-            indices_this_batch_this_process = [indices_this_batch[i] for i in self.dl._local_indices]
-            indices_for_this_batch_of_batches.extend(indices_this_batch_this_process)
+            global_offset = self.dl.scheduler.global_data_offset_by_step(bn)
+            offsets.append(global_offset)
+            local_indices_for_device = self.dl.local_data_indices_by_device_for_step(bn)
+
+            distinct_local_indices_this_batch = set()
+            for indices in local_indices_for_device.values():
+                for local_index in indices:
+                    distinct_local_indices_this_batch.add(local_index)
```

Comment:
What are instances where local_indices_for_device would have duplicates and it isn't a bug?

Reply:
tensor parallel, sequence parallel

Reply:
Got it - makes sense